<div align='center' ><font size='70'>Why LIR and LIRBase?</font></div>

<i class="fa fa-circle" aria-hidden="true"></i> <font size="4" color="red"><b>Long inverted repeat, long hpRNA and siRNA</b></font>  
>An inverted repeat is a single stranded nucleotide sequence followed by its reverse complement at the downstream. 
>The intervening sequence between the initial sequence and the reverse complement can be any length including zero. 
>When transcribed, **long inverted repeat can form long hairpin RNA genes (hpRNAs)**, which are much longer than typical animal or plant pre-miRNAs. 

>Henderson et al. reported the biogenesis of small interfering RNAs (siRNAs) from long inverted repeat in *Arabidopsis thaliana* for the first time (**Henderson et al. 2006 Nature genetics**).
>This siRNA biogenesis pathway was soon verified in *Drosophila* (**Czech et al. 2008 Nature**).
>In 2008, Okamura et al. systematically characterized the genes and mechanisms underlying the **biogenesis of 21-22-nucleotide siRNAs from long hpRNAs encoded by LIRs** in *Drosophila* (**Okamura et al. 2008 Nature**). 
>They found that Dicer-2, Hen1 and Argonaute 2 played vital roles in this siRNA biogenesis pathway. 
>This siRNA biogenesis pathway was further characterized in *Arabidopsis* soon (**Dunoyer et al. 2010 EMBO J**).

>**LIRs can act as functional genomic elements in eukaryotic genomes.**  
>A typical long inverted repeat and the small RNAs originated from the LIR analyzed utilizing LIRBase are demonstrated in the following image.  

<p class="aligncenter"><img src="LIR_rice.png" width="85%" height="70%" /></p>
<style>
.aligncenter {
    text-align: center;
}
</style>

<i class="fa fa-circle" aria-hidden="true"></i> <font size="4" color="red"><b>siRNA derived from long inverted repeats play important biological roles</b></font>  
>In 2018, Lin et al. identified two long hpRNAs in *Drosophila simulans*, which could be processed into 21-nt siRNAs (**Tao et al. 2007a PLOS Biology; Tao et al. 2007b PLOS Biology; Lin et al. 2018 Developmental Cell**). 
>These siRNAs could then repress the expression of the *Dox* and *MDox* genes which promotes X chromosome transmission by suppressing Y-bearing sperm.
>As a result, the **two long hpRNAs and the derived siRNAs are critical to the maintenance of balanced sex ratio** in the offsprings of *Drosophila simulans*.

>The biological functions of siRNAs derived from long inverted repeats in plants and animals were also reported in recent years.

>In **mouse**, siRNAs derived from LIRs were reported to regulate gene expression in oocytes (**Tam et al. 2008 Nature; Watanabe et al. 2008 Nature**).

>In **_Drosophila_**, another hpRNA and the derived siRNAs were reported to regulate testis gene expression and control male fertility (**Wen et al. 2015 Molecular Cell**).

>In **apple**, a long hpRNA and the generated siRNAs contributed to the resistance of apple to leaf spot disease (**Zhang et al. 2018 Plant Cell**).

>In **soybean**, a long hpRNA and the derived 22-nt siRNAs regulate the seed coat color of soybean (**Tuteja et al. 2009 Plant Cell; Cho et al. 2013 PLOS ONE; Jia et al. 2020 Plant Cell**).

>In **rice**, we previously found that several LIRs were present in one parental genome of an elite hybrid but were absent from the other parental genome (**Yao et al. 2020 Computational and Structural Biotechnology Journal**). As a result, siRNAs derived from the LIRs were detected and expressed in only one parental genome. The association between the LIRs and siRNAs were further detected and verified in an F2 population derived from a self-cross of the elite hybrid.

<p class="aligncenter"><img src="Lin2018.png" width="85%" height="70%" /></p>
<style>
.aligncenter {
    text-align: center;
}
</style>

<i class="fa fa-circle" aria-hidden="true"></i> <font size="4" color="red"><b>Comprehensive genome-wide identification of LIRs and long hpRNAs in eukaryotic genomes are urgently needed</b></font>  
>In 2013, **Axtell urgently called on the comprehensive genome-wide identification and annotation of long inverted repeats and long hpRNAs** (**Axtell et al. 2013 Annual Review of Plant Biology**).
>However, genome-wide identification and annotation of long inverted repeats were only conducted in very few organisms.
>None database or web server for annotation and analysis of long inverted repeats and long hpRNAs exist up to now.

>Using <a href="https://tandem.bu.edu/irf/irf.download.html" target="_blank">Inverted Repeats Finder (IRF)</a> (**Warburton et al. 2004 Genome Research**), we identified a total of 6,789,791 long inverted repeats in the whole genomes of 424 eukaryotes, including 297,317 LIRs in 77 invertebrate metazoa genomes, 1,902,296 LIRs in 142 plant genomes and 4,590,178 LIRs in 208 vertebrate genomes.
>We requested a minimum length of 400 nt for both arms of the long inverted repeat identified by IRF, to remove potential miniature inverted-repeat transposable element (MITE) or Alu element from the result of IRF.

<i class="fa fa-circle" aria-hidden="true"></i> <font size="4" color="red"><b>Nomenclature of a long inverted repeat in LIRBase</b></font>  
>**Each long inverted repeat has a unique identifier in LIRBase** determined by the species name and several features of the LIR including the chromosome ID, the start coordinate of the left arm, the end coordinate of the left arm, the start coordinate of the right arm, the end coordinate of the right arm.

>Please be noted that the sequence of a LIR in LIRBase is composed of the left arm sequence, the loop sequence, the right arm sequence, as well as two 200-bp sequences flanking the LIR (the left flanking sequence and the right flanking sequence). **The genomic coordinates of both arms of the LIR are reflected in the identifier of the LIR, while the flanking sequences are not denoted in the identifier of the LIR.** 


<p class="aligncenter"><img src="nomenclature.png" width="85%" height="70%" /></p>
<style>
.aligncenter {
    text-align: center;
}
</style>

<i class="fa fa-circle" aria-hidden="true"></i> <font size="4" color="red"><b>References</b></font>   
- Axtell et al. (2013), <a href="https://www.annualreviews.org/doi/abs/10.1146/annurev-arplant-050312-120043" target="_blank">Classification and Comparison of Small RNAs from Plants</a>, Annual Review of Plant Biology  
- Cho et al. (2013), <a href="https://doi.org/10.1371/journal.pone.0076954" target="_blank">The Transition from Primary siRNAs to Amplified Secondary siRNAs That Regulate Chalcone Synthase During Development of *Glycine max* Seed Coats</a>, PLoS ONE  
- Czech et al. (2008), <a href="https://www.nature.com/articles/nature07007" target="_blank">An endogenous small interfering RNA pathway in *Drosophila*</a>, Nature  
- Dunoyer et al. (2010), <a href="https://www.embopress.org/doi/full/10.1038/emboj.2010.65" target="_blank">An endogenous, systemic RNAi pathway in plants</a>, EMBO J (Note: This article had been retracted due to image irregularities, while the authors considered that the core conclusions of the published paper remain valid.)  
- Henderson et al. (2006), <a href="https://doi.org/10.1038/ng1804" target="_blank">Dissecting <i>Arabidopsis thaliana</i> DICER function in small RNA processing, gene silencing and DNA methylation patterning</a>, Nature genetics  
- Jia et al. (2020), <a href="http://www.plantcell.org/content/32/12/3662" target="_blank">Soybean DICER-LIKE2 Regulates Seed Coat Color via Production of Primary 22-Nucleotide Small Interfering RNAs from Long Inverted Repeats</a>, Plant Cell  
- Lin et al. (2018), <a href="https://doi.org/10.1016/j.devcel.2018.07.004" target="_blank">The hpRNA/RNAi Pathway Is Essential to Resolve Intragenomic Conflict in the *Drosophila* Male Germline</a>, Developmental Cell  
- Okamura et al. (2008), <a href="https://www.nature.com/articles/nature07015" target="_blank">The <i>Drosophila</i> hairpin RNA pathway generates endogenous short interfering RNAs</a>, Nature  
- Tam et al. (2008), <a href="https://doi.org/10.1038/nature06904" target="_blank">Pseudogene-derived small interfering RNAs regulate gene expression in mouse oocytes</a>, Nature  
- Tao et al. (2007), <a href="https://doi.org/10.1371/journal.pbio.0050292" target="_blank">A sex-ratio Meiotic Drive System in Drosophila simulans. I: An Autosomal Suppressor</a>, PLOS Biology  
- Tao et al. (2007), <a href="https://doi.org/10.1371/journal.pbio.0050293" target="_blank">A sex-ratio Meiotic Drive System in *Drosophila simulans*. II: An X-linked Distorter</a>, PLOS Biology  
- Tuteja et al. (2009), <a href="https://doi.org/10.1105/tpc.109.069856" target="_blank">Endogenous, Tissue-Specific Short Interfering RNAs Silence the Chalcone Synthase Gene Family in *Glycine max* Seed Coats</a>, Plant Cell  
- Warburton et al. (2004), <a href="https://genome.cshlp.org/content/14/10a/1861.abstract" target="_blank">Inverted Repeat Structure of the Human Genome: The X-Chromosome Contains a Preponderance of Large, Highly Homologous Inverted Repeats That Contain Testes Genes</a>, Genome Research  
- Watanabe et al. (2008), <a href="https://www.nature.com/articles/nature06908" target="_blank">Endogenous siRNAs from naturally formed dsRNAs regulate transcripts in mouse oocytes</a>, Nature  
- Wen et al. (2015), <a href="https://doi.org/10.1016/j.molcel.2014.11.025" target="_blank">Adaptive Regulation of Testis Gene Expression and Control of Male Fertility by the *Drosophila* Hairpin RNA Pathway</a>, Molecular Cell  
- Yao et al. (2020), <a href="https://doi.org/10.1016/j.csbj.2020.10.012" target="_blank">Features of sRNA biogenesis in rice revealed by genetic dissection of sRNA expression level</a>, Computational and Structural Biotechnology Journal  
- Zhang et al. (2018), <a href="http://www.plantcell.org/content/30/8/1924" target="_blank">A Single-Nucleotide Polymorphism in the Promoter of a Hairpin RNA Contributes to *Alternaria alternata* Leaf Spot Resistance in Apple (*Malus Ã— domestica*)</a>, Plant Cell  

<br><br><br>

